src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -8,111 +8,63 @@
 
-                libssl.SSL_CTX_set_quiet_shutdown(contextPtr, 1);
+                libssl.SSL_CTX_set_quiet_shutdown(innerContext, 1);
 
                 if (certHandle != null && certKeyHandle != null)
                 {
-                    SetSslCertificate(contextPtr, certHandle, certKeyHandle);
-                }
-
-                sslContext.sslPtr = libssl.SSL_new(contextPtr);
-
-                libssl.SSL_CTX_free(contextPtr);
-
-                if (IntPtr.Zero == sslContext.sslPtr)
-                {
-                    throw CreateSslException("Failed to create SSSL object from SSL context");
-                }
-
-                IntPtr memMethod = libcrypto.BIO_s_mem();
-
-                if (IntPtr.Zero == memMethod)
-                {
-                    throw CreateSslException("Failed to return memory BIO method function");
+                    SetSslCertificate(innerContext, certHandle, certKeyHandle);
                 }
 
-                sslContext.readBioPtr = libssl.BIO_new(memMethod);
-                sslContext.writeBioPtr = libssl.BIO_new(memMethod);
-
-                if ((IntPtr.Zero == sslContext.readBioPtr) || (IntPtr.Zero == sslContext.writeBioPtr))
-                {
-                    FreeBio(sslContext);
-                    throw CreateSslException("Failed to retun new BIO for a given method type");
-                }
-
-                if (isServer)
-                {
-                    libssl.SSL_set_accept_state(sslContext.sslPtr);
-                }
-                else
-                {
-                    libssl.SSL_set_connect_state(sslContext.sslPtr);
-                }
-
-                libssl.SSL_set_bio(sslContext.sslPtr, sslContext.readBioPtr, sslContext.writeBioPtr);
+                context = SafeSslHandle.Create(innerContext, isServer);
             }
-            catch
+
+            if ((null != context) && !context.IsInvalid)
             {
-                Disconnect(sslContext.sslPtr);
-                throw;
+                return context;
             }
 
-            IntPtr sslContextPtr = Marshal.AllocHGlobal(Marshal.SizeOf<SslContext>());
-            Marshal.StructureToPtr(sslContext, sslContextPtr, false);
-            return sslContextPtr;
+            if (null != context)
+            {
+                context.Dispose();
+            }
+            throw CreateSslException("Failed to create SSL object from SSL context");
         }
 
-        internal static bool DoSslHandshake(IntPtr sslContextPtr, IntPtr recvPtr, int recvCount, out IntPtr sendPtr, out int sendCount)
+        internal static bool DoSslHandshake(SafeSslHandle context, IntPtr recvPtr, int recvCount, out IntPtr sendPtr, out int sendCount)
         {
             sendPtr = IntPtr.Zero;
             sendCount = 0;
-            SslContext context = Marshal.PtrToStructure<SslContext>(sslContextPtr);
-            bool isServer = context.isServer;
+            bool isServer = context.IsServer;
```

Comment:
Nit: is the local actually needed?  It's used in just one place, which could just be:
```C#
if (retVal == 1 && context.IsServer)
```